apps/server/src/modules/user-login-migration/service/restart-user-login-migration-validation.service.spec.ts

```diff
@@ -0,0 +1,158 @@
+import { RestartUserLoginMigrationValidationService } from './restart-user-login-migration-validation.service';
+import { CommonUserLoginMigrationService } from './common-user-login-migration.service';
+
+describe('RestartUserLoginMigrationValidationService', () => {
+	let module: TestingModule;
+	let service: RestartUserLoginMigrationValidationService;
+
+	let commonUserLoginMigrationService: DeepMocked<CommonUserLoginMigrationService>;
+
+	beforeAll(async () => {
+		module = await Test.createTestingModule({
+			providers: [
+				RestartUserLoginMigrationValidationService,
+				{
+					provide: CommonUserLoginMigrationService,
+					useValue: createMock<CommonUserLoginMigrationService>(),
+				},
+			],
+		}).compile();
+
+		service = module.get(RestartUserLoginMigrationValidationService);
+		commonUserLoginMigrationService = module.get(CommonUserLoginMigrationService);
+	});
+
+	afterAll(async () => {
+		await module.close();
+	});
+
+	afterEach(() => {
+		jest.clearAllMocks();
+	});
+
+	describe('checkPreconditions', () => {
+		describe('when preconditions are met', () => {
+			const setup = () => {
+				const userId = 'userId';
+
+				const migration: UserLoginMigrationDO = new UserLoginMigrationDO({
+					schoolId: 'schoolId',
+					targetSystemId: 'targetSystemId',
+					startedAt: new Date('2022-12-17T03:24:00'),
+					closedAt: new Date('2023-12-17T03:24:00'),
+					finishedAt: new Date('2055-12-17T03:24:00'),
+				});
+
+				const school: SchoolDO = schoolDOFactory.buildWithId({ id: migration.id });
+
+				commonUserLoginMigrationService.ensurePermission.mockResolvedValue(Promise.resolve());
+				commonUserLoginMigrationService.findExistingUserLoginMigration.mockResolvedValue(migration);
+
+				return { userId, migration, schoolId: school.id as string };
+			};
+
+			it('should call ensurePermission', async () => {
+				const { userId, schoolId } = setup();
+
+				await service.checkPreconditions(userId, schoolId);
+
+				expect(commonUserLoginMigrationService.ensurePermission).toHaveBeenCalledWith(userId, schoolId);
+			});
+
+			it('should check if migration exists', async () => {
+				const { userId, schoolId } = setup();
+
+				await service.checkPreconditions(userId, schoolId);
+
+				expect(commonUserLoginMigrationService.findExistingUserLoginMigration).toHaveBeenCalledWith(schoolId);
+			});
+		});
+
+		describe('when migration could not be found', () => {
+			const setup = () => {
+				const userId = 'userId';
+
+				const school: SchoolDO = schoolDOFactory.buildWithId();
+
+				commonUserLoginMigrationService.ensurePermission.mockResolvedValue(Promise.resolve());
+				commonUserLoginMigrationService.findExistingUserLoginMigration.mockResolvedValue(null);
+
+				return { userId, schoolId: school.id as string };
+			};
+
+			it('should throw RestartUserLoginMigrationError ', async () => {
+				const { userId, schoolId } = setup();
+
+				await expect(service.checkPreconditions(userId, schoolId)).rejects.toThrow(
```

Comment:
Here you can seperate the function and the expection for more readability like in the start pr. Please do that in all functions